packages/core/src/services/documentVersions/materializeAtCommit.ts

```diff
@@ -1,28 +1,30 @@
-import { uniqBy } from 'lodash-es'
-
+import { database } from '$core/client'
 import { HEAD_COMMIT } from '$core/constants'
-import {
-  getDocumentsAtCommit,
-  listdocumentSnapshots,
-  listStagedDocuments,
-} from '$core/data-access'
+import { findCommit } from '$core/data-access'
+import { Result } from '$core/lib'
+import { NotFoundError } from '$core/lib/errors'
+import { commits, documentVersions } from '$core/schema'
+import { and, desc, eq, isNotNull, lte } from 'drizzle-orm'
 
 export async function materializeDocumentsAtCommit({
   commitUuid = HEAD_COMMIT,
-  staged = true,
 }: {
   commitUuid: string
-  staged: boolean
 }) {
-  if (commitUuid === HEAD_COMMIT) {
-    const snapshots = (await listdocumentSnapshots()).map(
-      (snap) => snap.document_versions,
+  const commit = await findCommit({ uuid: commitUuid })
+  if (!commit || !commit.mergedAt) {
+    return Result.error(new NotFoundError('Commit not found'))
+  }
+
+  const docsInCommits = await database
+    .selectDistinct()
+    .from(documentVersions)
+    .innerJoin(commits, eq(commits.id, documentVersions.commitId))
+    .where(
+      and(isNotNull(commits.mergedAt), lte(commits.mergedAt, commit.mergedAt!)),
     )
-    if (!staged) return snapshots
+    .groupBy(documentVersions.documentUuid, documentVersions.id, commits.id)
+    .orderBy(desc(commits.mergedAt))
 
-    const versions = await listStagedDocuments()
-    return uniqBy([...versions, ...snapshots], (doc) => doc.documentUuid)
-  } else {
-    return await getDocumentsAtCommit(commitUuid)
-  }
+  return Result.ok(docsInCommits.map((doc) => doc.document_versions))
```

Comment:
this query gets you all documents in a main branch commit, no staged documents. Useful for production use cases where you wanna target only main branch commits.